File: src/main/docker.js
const { app, ipcMain } = require('electron');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const { Client } = require('ssh2');

let mainWindow = null;
const connections = new Map(); // hostId -> ssh2.Client

// ========== Config Persistence ==========

function getConfigPath() {
  return path.join(app.getPath('userData'), 'docker-hosts.json');
}

function loadHosts() {
  try {
    return JSON.parse(fs.readFileSync(getConfigPath(), 'utf8'));
  } catch {
    return [];
  }
}

function saveHosts(hosts) {
  fs.writeFileSync(getConfigPath(), JSON.stringify(hosts, null, 2));
}

// ========== SSH Connection Pool ==========

function connectToHost(hostConfig) {
  return new Promise((resolve, reject) => {
    // Fermer connexion existante si présente
    if (connections.has(hostConfig.id)) {
      try { connections.get(hostConfig.id).end(); } catch {}
      connections.delete(hostConfig.id);
    }

    const conn = new Client();
    const config = {
      host: hostConfig.hostname,
      port: hostConfig.port || 22,
      username: hostConfig.username,
      readyTimeout: 10000,
    };

    if (hostConfig.authType === 'key' && hostConfig.privateKeyPath) {
      try {
        config.privateKey = fs.readFileSync(hostConfig.privateKeyPath);
        if (hostConfig.passphrase) config.passphrase = hostConfig.passphrase;
      } catch (err) {
        return reject(new Error(`Cannot read SSH key: ${err.message}`));
      }
    } else {
      config.password = hostConfig.password;
    }

    conn.on('ready', () => {
      connections.set(hostConfig.id, conn);
      resolve(conn);
    });

    conn.on('error', (err) => {
      connections.delete(hostConfig.id);
      reject(err);
    });

    conn.on('close', () => {
      connections.delete(hostConfig.id);
    });

    conn.connect(config);
  });
}

function disconnectHost(hostId) {
  const conn = connections.get(hostId);
  if (conn) {
    try { conn.end(); } catch {}
    connections.delete(hostId);
  }
}

function disconnectAll() {
  for (const [, conn] of connections) {
    try { conn.end(); } catch {}
  }
  connections.clear();
}

// ========== SSH Command Execution ==========

function execSSH(conn, command, timeout = 15000) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error('SSH command timeout'));
    }, timeout);

    conn.exec(command, (err, stream) => {
      if (err) {
        clearTimeout(timer);
        return reject(err);
      }

      let stdout = '';
      let stderr = '';

      stream.on('data', (data) => { stdout += data.toString(); });
      stream.stderr.on('data', (data) => { stderr += data.toString(); });
      stream.on('close', (code) => {
        clearTimeout(timer);
        if (code !== 0 && !stdout) {
          reject(new Error(stderr || `Exit code ${code}`));
        } else {
          resolve({ stdout, stderr, code });
        }
      });
    });
  });
}

// ========== Security ==========

function isValidContainerId(id) {
  return /^[a-zA-Z0-9][a-zA-Z0-9_.-]*$/.test(id);
}

// ========== Docker CLI Wrappers ==========

async function listContainers(conn) {
  const { stdout } = await execSSH(conn,
    "docker ps -a --format '{{json .}}'"
  );
  if (!stdout.trim()) return [];
  return stdout.trim().split('\n').filter(Boolean).map(line => {
    const c = JSON.parse(line);
    return {
      id: c.ID,
      name: c.Names,
      image: c.Image,
      status: c.Status,
      state: c.State,
      ports: c.Ports,
      createdAt: c.CreatedAt,
      runningFor: c.RunningFor,
    };
  });
}

async function getContainerStats(conn) {
  const { stdout } = await execSSH(conn,
    "docker stats --no-stream --format '{{json .}}'",
    30000
  );
  if (!stdout.trim()) return [];
  return stdout.trim().split('\n').filter(Boolean).map(line => {
    const s = JSON.parse(line);
    return {
      id: s.ID,
      name: s.Name,
      cpuPerc: s.CPUPerc,
      memUsage: s.MemUsage,
      memPerc: s.MemPerc,
      netIO: s.NetIO,
      blockIO: s.BlockIO,
    };
  });
}

async function inspectContainer(conn, containerId) {
  if (!isValidContainerId(containerId)) throw new Error('Invalid container ID');
  const { stdout } = await execSSH(conn, `docker inspect ${containerId}`);
  const data = JSON.parse(stdout);
  if (!data || !data[0]) throw new Error('Container not found');
  const c = data[0];
  return {
    id: c.Id,
    name: (c.Name || '').replace(/^\//, ''),
    image: c.Config?.Image || '',
    env: c.Config?.Env || [],
    ports: c.NetworkSettings?.Ports || {},
    volumes: (c.Mounts || []).map(m => ({
      source: m.Source,
      destination: m.Destination,
      mode: m.Mode,
      type: m.Type,
    })),
    restartPolicy: c.HostConfig?.RestartPolicy?.Name || 'no',
    state: c.State?.Status,
    startedAt: c.State?.StartedAt,
    finishedAt: c.State?.FinishedAt,
    created: c.Created,
    composeProject: c.Config?.Labels?.['com.docker.compose.project'] || null,
    composeService: c.Config?.Labels?.['com.docker.compose.service'] || null,
    composeWorkingDir: c.Config?.Labels?.['com.docker.compose.project.working_dir'] || null,
  };
}

async function getContainerLogs(conn, containerId, tail = 50) {
  if (!isValidContainerId(containerId)) throw new Error('Invalid container ID');
  const { stdout } = await execSSH(conn,
    `docker logs --tail ${parseInt(tail, 10)} --timestamps ${containerId} 2>&1`,
    30000
  );
  return stdout;
}

async function containerAction(conn, containerId, action) {
  const validActions = ['start', 'stop', 'restart'];
  if (!validActions.includes(action)) throw new Error('Invalid action');
  if (!isValidContainerId(containerId)) throw new Error('Invalid container ID');
  await execSSH(conn, `docker ${action} ${containerId}`, 30000);
}

async function updateContainer(conn, containerId) {
  if (!isValidContainerId(containerId)) throw new Error('Invalid container ID');

  // Récupérer les infos du conteneur (image + labels compose)
  const { stdout } = await execSSH(conn, `docker inspect ${containerId}`);
  const data = JSON.parse(stdout);
  if (!data || !data[0]) throw new Error('Container not found');
  const c = data[0];

  const image = c.Config?.Image || '';
  const composeService = c.Config?.Labels?.['com.docker.compose.service'] || null;
  const composeWorkingDir = c.Config?.Labels?.['com.docker.compose.project.working_dir'] || null;

  if (!image) throw new Error('No image found for container');

  if (composeService && composeWorkingDir) {
    // Docker Compose : pull + recreate
    await execSSH(conn,
      `cd ${composeWorkingDir} && docker compose pull ${composeService} && docker compose up -d ${composeService}`,
      120000
    );
    return { compose: true, pulled: true };
  } else {
    // Standalone : pull uniquement
    await execSSH(conn, `docker pull ${image}`, 120000);
    return { compose: false, pulled: true };
  }
}

// ========== IPC Handlers ==========

function initDocker(win) {
  mainWindow = win;
}

function registerDockerIpc() {
  // Host management
  ipcMain.handle('docker-get-hosts', () => {
    return loadHosts();
  });

  ipcMain.handle('docker-save-hosts', (_event, hosts) => {
    saveHosts(hosts);
    return { success: true };
  });

  // Connection
  ipcMain.handle('docker-connect', async (_event, hostId) => {
    const hosts = loadHosts();
    const host = hosts.find(h => h.id === hostId);
    if (!host) return { success: false, error: 'Host not found' };
    try {
      await connectToHost(host);
      return { success: true };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-disconnect', (_event, hostId) => {
    disconnectHost(hostId);
    return { success: true };
  });

  ipcMain.handle('docker-test-connection', async (_event, hostConfig) => {
    const conn = new Client();
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        try { conn.end(); } catch {}
        resolve({ success: false, error: 'Connection timeout' });
      }, 10000);

      const config = {
        host: hostConfig.hostname,
        port: hostConfig.port || 22,
        username: hostConfig.username,
        readyTimeout: 10000,
      };

      if (hostConfig.authType === 'key' && hostConfig.privateKeyPath) {
        try {
          config.privateKey = fs.readFileSync(hostConfig.privateKeyPath);
          if (hostConfig.passphrase) config.passphrase = hostConfig.passphrase;
        } catch (err) {
          clearTimeout(timer);
          return resolve({ success: false, error: `Cannot read SSH key: ${err.message}` });
        }
      } else {
        config.password = hostConfig.password;
      }

      conn.on('ready', () => {
        clearTimeout(timer);
        // Vérifier que Docker est installé
        conn.exec('docker --version', (err, stream) => {
          if (err) {
            conn.end();
            return resolve({ success: false, error: err.message });
          }
          let out = '';
          stream.on('data', (data) => { out += data.toString(); });
          stream.on('close', (code) => {
            conn.end();
            if (code !== 0) {
              resolve({ success: false, error: 'Docker is not installed on this host' });
            } else {
              resolve({ success: true, dockerVersion: out.trim() });
            }
          });
        });
      });

      conn.on('error', (err) => {
        clearTimeout(timer);
        if (err.level === 'client-authentication') {
          resolve({ success: false, error: 'SSH authentication failed' });
        } else {
          resolve({ success: false, error: err.message });
        }
      });

      conn.connect(config);
    });
  });

  // Container operations
  ipcMain.handle('docker-list-containers', async (_event, hostId) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      const containers = await listContainers(conn);
      return { success: true, data: containers };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-get-stats', async (_event, hostId) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      const stats = await getContainerStats(conn);
      return { success: true, data: stats };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-inspect', async (_event, hostId, containerId) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      const info = await inspectContainer(conn, containerId);
      return { success: true, data: info };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-logs', async (_event, hostId, containerId, tail) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      const logs = await getContainerLogs(conn, containerId, tail || 50);
      return { success: true, data: logs };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-action', async (_event, hostId, containerId, action) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      await containerAction(conn, containerId, action);
      return { success: true };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  ipcMain.handle('docker-update-container', async (_event, hostId, containerId) => {
    const conn = connections.get(hostId);
    if (!conn) return { success: false, error: 'Not connected' };
    try {
      const result = await updateContainer(conn, containerId);
      return { success: true, data: result };
    } catch (err) {
      return { success: false, error: err.message };
    }
  });

  // SSH key file picker
  ipcMain.handle('docker-select-ssh-key', async () => {
    const { dialog } = require('electron');
    const os = require('os');
    const result = await dialog.showOpenDialog(mainWindow, {
      title: 'Select SSH Private Key',
      properties: ['openFile'],
      filters: [{ name: 'All Files', extensions: ['*'] }],
      defaultPath: path.join(os.homedir(), '.ssh'),
    });
    if (result.canceled) return { success: false };
    return { success: true, path: result.filePaths[0] };
  });
}

function stopDocker() {
  disconnectAll();
}

module.exports = { initDocker, registerDockerIpc, stopDocker };

File: src-tauri/src/commands/docker.rs
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use tauri::Manager;
use tauri_plugin_dialog::DialogExt;

// SSH connection pool (global state)
use once_cell::sync::Lazy;
static CONNECTIONS: Lazy<Arc<Mutex<HashMap<String, Arc<russh::client::Handle<SshHandler>>>>>> =
    Lazy::new(|| Arc::new(Mutex::new(HashMap::new())));

// Minimal SSH handler
struct SshHandler;

#[async_trait::async_trait]
impl russh::client::Handler for SshHandler {
    type Error = russh::Error;

    async fn check_server_key(
        &mut self,
        _server_public_key: &russh_keys::key::PublicKey,
    ) -> Result<bool, Self::Error> {
        Ok(true) // Accept all host keys (like ssh2 in Electron version)
    }
}

/// Execute command over SSH and return stdout
async fn exec_ssh(
    handle: &russh::client::Handle<SshHandler>,
    command: &str,
    timeout_secs: u64,
) -> Result<String, String> {
    let mut channel = handle.channel_open_session().await.map_err(|e| e.to_string())?;
    channel.exec(true, command).await.map_err(|e| e.to_string())?;

    let mut stdout = Vec::new();
    let timeout = tokio::time::Duration::from_secs(timeout_secs);

    loop {
        match tokio::time::timeout(timeout, channel.wait()).await {
            Ok(Some(msg)) => {
                match msg {
                    russh::ChannelMsg::Data { data } => {
                        stdout.extend_from_slice(&data);
                    }
                    russh::ChannelMsg::ExtendedData { data, .. } => {
                        // stderr - also capture for docker logs 2>&1
                        stdout.extend_from_slice(&data);
                    }
                    russh::ChannelMsg::Eof | russh::ChannelMsg::Close => break,
                    russh::ChannelMsg::ExitStatus { .. } => {}
                    _ => {}
                }
            }
            Ok(None) => break,
            Err(_) => return Err("SSH command timeout".into()),
        }
    }

    Ok(String::from_utf8_lossy(&stdout).to_string())
}

fn is_valid_container_id(id: &str) -> bool {
    !id.is_empty() && id.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '.' || c == '-')
}

// ========== IPC Handlers ==========

#[tauri::command]
pub async fn docker_get_hosts(app: tauri::AppHandle) -> Result<Value, String> {
    let path = app.path().app_data_dir().unwrap().join("docker-hosts.json");
    if !path.exists() {
        return Ok(serde_json::json!([]));
    }
    let content = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let data: Value = serde_json::from_str(&content).map_err(|e| e.to_string())?;
    Ok(data)
}

#[tauri::command]
pub async fn docker_save_hosts(app: tauri::AppHandle, hosts: Value) -> Result<Value, String> {
    let path = app.path().app_data_dir().unwrap().join("docker-hosts.json");
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    std::fs::write(&path, serde_json::to_string_pretty(&hosts).unwrap())
        .map_err(|e| e.to_string())?;
    Ok(serde_json::json!({ "success": true }))
}

#[tauri::command]
pub async fn docker_connect(app: tauri::AppHandle, host_id: String) -> Result<Value, String> {
    let path = app.path().app_data_dir().unwrap().join("docker-hosts.json");
    let hosts: Vec<Value> = if path.exists() {
        let content = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
        serde_json::from_str(&content).map_err(|e| e.to_string())?
    } else {
        return Err("No hosts configured".into());
    };

    let host = hosts.iter()
        .find(|h| h["id"].as_str() == Some(&host_id))
        .ok_or("Host not found")?;

    let hostname = host["hostname"].as_str().unwrap_or("localhost").to_string();
    let port = host["port"].as_u64().unwrap_or(22) as u16;
    let username = host["username"].as_str().unwrap_or("root").to_string();
    let auth_type = host["authType"].as_str().unwrap_or("password");
    let password = host["password"].as_str().unwrap_or("").to_string();
    let key_path = host["privateKeyPath"].as_str().unwrap_or("").to_string();
    let passphrase = host["passphrase"].as_str().map(|s| s.to_string());

    // Disconnect existing
    {
        let mut conns = CONNECTIONS.lock().await;
        conns.remove(&host_id);
    }

    let config = Arc::new(russh::client::Config::default());
    let handler = SshHandler;

    let mut session = russh::client::connect(config, (hostname.as_str(), port), handler)
        .await
        .map_err(|e| format!("SSH connection failed: {}", e))?;

    // Authenticate
    let auth_ok = if auth_type == "key" && !key_path.is_empty() {
        let key_data = std::fs::read_to_string(&key_path)
            .map_err(|e| format!("Cannot read SSH key: {}", e))?;
        let key = if let Some(ref pass) = passphrase {
            russh_keys::decode_secret_key(&key_data, Some(pass))
        } else {
            russh_keys::decode_secret_key(&key_data, None)
        }
        .map_err(|e| format!("Invalid SSH key: {}", e))?;

        session.authenticate_publickey(&username, Arc::new(key))
            .await
            .map_err(|e| format!("SSH auth failed: {}", e))?
    } else {
        session.authenticate_password(&username, &password)
            .await
            .map_err(|e| format!("SSH auth failed: {}", e))?
    };

    if !auth_ok {
        return Err("SSH authentication failed".into());
    }

    let handle = Arc::new(session);
    CONNECTIONS.lock().await.insert(host_id, handle);

    Ok(serde_json::json!({ "success": true }))
}

#[tauri::command]
pub async fn docker_disconnect(host_id: String) -> Result<Value, String> {
    CONNECTIONS.lock().await.remove(&host_id);
    Ok(serde_json::json!({ "success": true }))
}

#[tauri::command]
pub async fn docker_test_connection(config: Value) -> Result<Value, String> {
    let hostname = config["hostname"].as_str().unwrap_or("localhost").to_string();
    let port = config["port"].as_u64().unwrap_or(22) as u16;
    let username = config["username"].as_str().unwrap_or("root").to_string();
    let auth_type = config["authType"].as_str().unwrap_or("password");
    let password = config["password"].as_str().unwrap_or("").to_string();
    let key_path = config["privateKeyPath"].as_str().unwrap_or("").to_string();
    let passphrase = config["passphrase"].as_str().map(|s| s.to_string());

    let ssh_config = Arc::new(russh::client::Config {
        ..Default::default()
    });

    let mut session = match tokio::time::timeout(
        tokio::time::Duration::from_secs(10),
        russh::client::connect(ssh_config, (hostname.as_str(), port), SshHandler),
    ).await {
        Ok(Ok(s)) => s,
        Ok(Err(e)) => return Ok(serde_json::json!({ "success": false, "error": e.to_string() })),
        Err(_) => return Ok(serde_json::json!({ "success": false, "error": "Connection timeout" })),
    };

    let auth_ok = if auth_type == "key" && !key_path.is_empty() {
        let key_data = match std::fs::read_to_string(&key_path) {
            Ok(d) => d,
            Err(e) => return Ok(serde_json::json!({ "success": false, "error": format!("Cannot read SSH key: {}", e) })),
        };
        let key = match if let Some(ref pass) = passphrase {
            russh_keys::decode_secret_key(&key_data, Some(pass))
        } else {
            russh_keys::decode_secret_key(&key_data, None)
        } {
            Ok(k) => k,
            Err(e) => return Ok(serde_json::json!({ "success": false, "error": format!("Invalid SSH key: {}", e) })),
        };
        session.authenticate_publickey(&username, Arc::new(key)).await.unwrap_or(false)
    } else {
        session.authenticate_password(&username, &password).await.unwrap_or(false)
    };

    if !auth_ok {
        return Ok(serde_json::json!({ "success": false, "error": "SSH authentication failed" }));
    }

    // Check docker is installed
    let handle = Arc::new(session);
    match exec_ssh(&handle, "docker --version", 10).await {
        Ok(version) => Ok(serde_json::json!({ "success": true, "dockerVersion": version.trim() })),
        Err(_) => Ok(serde_json::json!({ "success": false, "error": "Docker is not installed on this host" })),
    }
}

#[tauri::command]
pub async fn docker_list_containers(host_id: String) -> Result<Value, String> {
    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;

    let stdout = exec_ssh(handle, "docker ps -a --format '{{json .}}'", 15).await?;
    let containers: Vec<Value> = stdout.lines()
        .filter(|l| !l.is_empty())
        .filter_map(|l| serde_json::from_str::<Value>(l).ok())
        .map(|c| serde_json::json!({
            "id": c["ID"],
            "name": c["Names"],
            "image": c["Image"],
            "status": c["Status"],
            "state": c["State"],
            "ports": c["Ports"],
            "createdAt": c["CreatedAt"],
            "runningFor": c["RunningFor"],
        }))
        .collect();

    Ok(serde_json::json!({ "success": true, "data": containers }))
}

#[tauri::command]
pub async fn docker_get_stats(host_id: String) -> Result<Value, String> {
    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;

    let stdout = exec_ssh(handle, "docker stats --no-stream --format '{{json .}}'", 30).await?;
    let stats: Vec<Value> = stdout.lines()
        .filter(|l| !l.is_empty())
        .filter_map(|l| serde_json::from_str::<Value>(l).ok())
        .map(|s| serde_json::json!({
            "id": s["ID"],
            "name": s["Name"],
            "cpuPerc": s["CPUPerc"],
            "memUsage": s["MemUsage"],
            "memPerc": s["MemPerc"],
            "netIO": s["NetIO"],
            "blockIO": s["BlockIO"],
        }))
        .collect();

    Ok(serde_json::json!({ "success": true, "data": stats }))
}

#[tauri::command]
pub async fn docker_inspect(host_id: String, container_id: String) -> Result<Value, String> {
    if !is_valid_container_id(&container_id) {
        return Err("Invalid container ID".into());
    }

    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;

    let stdout = exec_ssh(handle, &format!("docker inspect {}", container_id), 15).await?;
    let data: Vec<Value> = serde_json::from_str(&stdout).map_err(|e| e.to_string())?;
    let c = data.first().ok_or("Container not found")?;

    let volumes: Vec<Value> = c["Mounts"].as_array()
        .map(|mounts| mounts.iter().map(|m| serde_json::json!({
            "source": m["Source"],
            "destination": m["Destination"],
            "mode": m["Mode"],
            "type": m["Type"],
        })).collect())
        .unwrap_or_default();

    Ok(serde_json::json!({
        "success": true,
        "data": {
            "id": c["Id"],
            "name": c["Name"].as_str().unwrap_or("").trim_start_matches('/'),
            "image": c["Config"]["Image"],
            "env": c["Config"]["Env"],
            "ports": c["NetworkSettings"]["Ports"],
            "volumes": volumes,
            "restartPolicy": c["HostConfig"]["RestartPolicy"]["Name"],
            "state": c["State"]["Status"],
            "startedAt": c["State"]["StartedAt"],
            "finishedAt": c["State"]["FinishedAt"],
            "created": c["Created"],
            "composeProject": c["Config"]["Labels"]["com.docker.compose.project"],
            "composeService": c["Config"]["Labels"]["com.docker.compose.service"],
            "composeWorkingDir": c["Config"]["Labels"]["com.docker.compose.project.working_dir"],
        }
    }))
}

#[tauri::command]
pub async fn docker_logs(host_id: String, container_id: String, tail: Option<i32>) -> Result<Value, String> {
    if !is_valid_container_id(&container_id) {
        return Err("Invalid container ID".into());
    }

    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;
    let tail_n = tail.unwrap_or(50);

    let stdout = exec_ssh(
        handle,
        &format!("docker logs --tail {} --timestamps {} 2>&1", tail_n, container_id),
        30,
    ).await?;

    Ok(serde_json::json!({ "success": true, "data": stdout }))
}

#[tauri::command]
pub async fn docker_action(host_id: String, container_id: String, action: String) -> Result<Value, String> {
    if !["start", "stop", "restart"].contains(&action.as_str()) {
        return Err("Invalid action".into());
    }
    if !is_valid_container_id(&container_id) {
        return Err("Invalid container ID".into());
    }

    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;

    exec_ssh(handle, &format!("docker {} {}", action, container_id), 30).await?;
    Ok(serde_json::json!({ "success": true }))
}

#[tauri::command]
pub async fn docker_update_container(host_id: String, container_id: String) -> Result<Value, String> {
    if !is_valid_container_id(&container_id) {
        return Err("Invalid container ID".into());
    }

    let conns = CONNECTIONS.lock().await;
    let handle = conns.get(&host_id).ok_or("Not connected")?;

    // Get container info for image and compose labels
    let stdout = exec_ssh(handle, &format!("docker inspect {}", container_id), 15).await?;
    let data: Vec<Value> = serde_json::from_str(&stdout).map_err(|e| e.to_string())?;
    let c = data.first().ok_or("Container not found")?;

    let image = c["Config"]["Image"].as_str().unwrap_or("");
    let compose_service = c["Config"]["Labels"]["com.docker.compose.service"].as_str();
    let compose_dir = c["Config"]["Labels"]["com.docker.compose.project.working_dir"].as_str();

    if image.is_empty() {
        return Err("No image found for container".into());
    }

    if let (Some(service), Some(dir)) = (compose_service, compose_dir) {
        // Docker Compose: pull + recreate
        exec_ssh(
            handle,
            &format!("cd {} && docker compose pull {} && docker compose up -d {}", dir, service, service),
            120,
        ).await?;
        Ok(serde_json::json!({ "success": true, "data": { "compose": true, "pulled": true } }))
    } else {
        // Standalone: pull only
        exec_ssh(handle, &format!("docker pull {}", image), 120).await?;
        Ok(serde_json::json!({ "success": true, "data": { "compose": false, "pulled": true } }))
    }
}

#[tauri::command]
pub async fn docker_select_ssh_key(app: tauri::AppHandle) -> Result<Value, String> {
    let home = dirs::home_dir().unwrap_or_default();
    let default_path = home.join(".ssh");

    let file = app.dialog()
        .file()
        .set_directory(&default_path)
        .set_title("Select SSH Private Key")
        .blocking_pick_file();

    match file {
        Some(path) => {
            let key_path = path.as_path().map(|p| p.to_string_lossy().to_string()).unwrap_or_default();
            Ok(serde_json::json!({
                "success": true,
                "path": key_path,
            }))
        }
        None => Ok(serde_json::json!({ "success": false })),
    }
}
